Feat / Add connector for Gate.io

TheHolyRoger opened the pull request:
**Before submitting this PR, please make sure**:

- [X] Your code builds clean without any errors or warnings
- [X] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
Adds a connector for Gate.io!

PtrckM commented:
- inconsistent order creation when hanging order is enabled
![image](https://user-images.githubusercontent.com/54516858/120879137-de2abe00-c5f3-11eb-8460-7c5324f98bc9.png)
[gate_io.zip](https://github.com/CoinAlpha/hummingbot/files/6601964/gate_io.zip)

TheHolyRoger commented:
@PtrckM thank you! Looking at this now

TheHolyRoger commented:
@PtrckM could you see how this works with latest changes please? Fixed an issue with order status update.

Also bear in mind that there will be slight delays with balance updates after order creation/cancel due to Gate.io websocket

rxlxrxsx commented:
1. There are instances wherein the bot takes time before placing orders after starting the strategy.

<img width="697" alt="Screen Shot 2021-06-06 at 2 32 17 PM" src="https://user-images.githubusercontent.com/64377055/120917446-f24ce900-c6e1-11eb-89e8-a63c893c334e.png">

2. There are also instances that the available balance takes time to update, in this screenshot you'll see that the available balance is still not updating even when there are open orders for 36 seconds already.

<img width="546" alt="Screen Shot 2021-06-06 at 1 51 45 PM" src="https://user-images.githubusercontent.com/64377055/120917482-26280e80-c6e2-11eb-9939-6e3b87faad00.png">

3. Hanging orders refreshed by max order age is not recreated due to the delay in balance update.

```
2021-06-06 14:53:41,196 - 11561 - hummingbot.connector.exchange.gate_io.gate_io_exchange - NETWORK - Error submitting SELL LIMIT order to gate_io for 34.94 DOGE-USDT 0.373463 - None.
Traceback (most recent call last):
  File "/Users/reiniersalazar/hummingbot/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 450, in _create_order
    order_result = await self._api_request("POST", Constants.ENDPOINT["ORDER_CREATE"], api_params, True)
  File "/Users/reiniersalazar/hummingbot/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 353, in _api_request
    raise GateIoAPIError(parsed_response)
hummingbot.connector.exchange.gate_io.gate_io_utils.GateIoAPIError: {'label': 'BALANCE_NOT_ENOUGH', 'message': 'Not enough balance'}
2021-06-06 14:53:41,199 - 11561 - hummingbot.connector.exchange.gate_io.gate_io_exchange - WARNING - Error submitting order to gate_io - None.
```

<img width="1422" alt="Screen Shot 2021-06-06 at 2 39 37 PM" src="https://user-images.githubusercontent.com/64377055/120917547-71422180-c6e2-11eb-80d5-c1a74ac9bd26.png">

Logs and config: [gate_io_test.zip](https://github.com/CoinAlpha/hummingbot/files/6603837/gate_io_test.zip)

TheHolyRoger commented:
Thanks @rxlxrxsx I've done a couple of changes that should help this. Will still be a bit of a delay in balance updates due to the gate.io websocket implementation

rxlxrxsx commented:
- Confirming that this issue is now fixed.

> There are also instances that the available balance takes time to update, in this screenshot you'll see that the available balance is still not updating even when there are open orders for 36 seconds already.

- I can see that this is still happening on the latest commit although not that long compared to the previous commit.

> There are instances wherein the bot takes time before placing orders after starting the strategy.

<img width="689" alt="Screen Shot 2021-06-07 at 11 47 27 AM" src="https://user-images.githubusercontent.com/64377055/120973572-af524a80-c7a1-11eb-87fa-2b7f63b20168.png">

- This is still happening on the latest commit.

> Hanging orders refreshed by max order age is not recreated due to the delay in balance update.

<img width="1423" alt="Screen Shot 2021-06-07 at 12 17 52 PM" src="https://user-images.githubusercontent.com/64377055/120973768-f04a5f00-c7a1-11eb-9b83-f52e0277fd1c.png">

- There are instances that the bot skips an order cycle and won't create orders during those period.

<img width="1421" alt="Screen Shot 2021-06-07 at 12 09 11 PM" src="https://user-images.githubusercontent.com/64377055/120973869-11ab4b00-c7a2-11eb-9f17-3b0c216b0951.png">

Logs and config: [gate_io_updated.zip](https://github.com/CoinAlpha/hummingbot/files/6606595/gate_io_updated.zip)

TheHolyRoger commented:
@rxlxrxsx thanks man, think I've addressed these bits some more. the Hanging orders balance issue is difficult to reproduce, I've only seen it a couple of times... Have you got a config that makes it happen more frequently?

rxlxrxsx commented:
Currently I was able to replicate it using the config file attached above but I'll let you know immediately if I'll be able to replicate it more frequently using a different config.

Nullably commented:
@TheHolyRoger We have identified the cause for a delay in connector ready issue. This may apply to gate.io connector as well. See https://github.com/CoinAlpha/hummingbot/pull/3474

TheHolyRoger commented:
@Nullably Thanks! applied that now. Also regarding the issue rxl is experiencing above, the hanging order refresh cycle in PMM does seem to fire cancel + re-create almost simultaneously and prone to race conditions... What are your thoughts?

Any way I can solve this better in the connector since we don't get balance updates via websockets?

rxlxrxsx commented:
- This issue is now fixed.

> There are instances wherein the bot takes time before placing orders after starting the strategy.

- I was able to replicate this again using the config attached below.

> Hanging orders refreshed by max order age is not recreated due to the delay in balance update.

<img width="1421" alt="Screen Shot 2021-06-08 at 5 10 33 PM" src="https://user-images.githubusercontent.com/64377055/121167620-ae461980-c884-11eb-95ce-7ebd1f9e6932.png">

- This is still happening on the latest commit.

> There are instances that the bot skips an order cycle and won't create orders during those period.

<img width="1424" alt="Screen Shot 2021-06-08 at 5 07 36 PM" src="https://user-images.githubusercontent.com/64377055/121166751-ebf67280-c883-11eb-9bbd-b5f29d5c26eb.png">

- Inconsistent order creation is happening again, even without a hanging order currently in place

<img width="1422" alt="Screen Shot 2021-06-08 at 5 54 15 PM" src="https://user-images.githubusercontent.com/64377055/121167245-560f1780-c884-11eb-8764-98538c1350e0.png">

Logs and config: [gate_io_files.zip](https://github.com/CoinAlpha/hummingbot/files/6615629/gate_io_files.zip)

TheHolyRoger commented:
@rxlxrxsx thank you! Waiting for @Nullably thoughts on how best to handle balance update lag... 

If the strategies wait for the order cancelled event before proceeding, I could remove the triggering of that event during `execute_cancel` and instead fire the cancel event only when the websocket order status message is received?

Nullably commented:
@TheHolyRoger 
> Any way I can solve this better in the connector since we don't get balance updates via websockets?

You can set _real_time_balance_update to False (in init), see hummingbot/connector/connector_base.pyx
This will turn on an internal accounting for balances.
You will need to set `_in_flight_orders_snapshot` and `_in_flight_orders_snapshot_timestamp` on `_update_balance` as well. See `hummingbot/connector/exchange/coinbase_pro/coinbase_pro_exchange.pyx` for an example.

TheHolyRoger commented:
@Nullably thank you! Didn't know about that one. I've also added the changes for coinzoom here: https://github.com/CoinAlpha/hummingbot/pull/3520

@rxlxrxsx could you have another go and see if it's better now please? :) Seems better for me

TheHolyRoger commented:
@Nullably Damn, I'm still getting these occasionally rejected orders due to balance not updating quick enough...

I'm looking at the PMM code for the `c_aged_order_refresh` logic - there doesn't seem to be any mechanism to "wait" for the balance to be updated before recreating hanging orders.

The order is cancelled, added to the current proposal and then immediately re-created in the same cycle, bypassing available balance checks...

I don't see any way to help this in the connector itself... any thoughts?

rxlxrxsx commented:
- These issues are now fixed.

> There are instances that the bot skips an order cycle and won't create orders during those period.

> Inconsistent order creation is happening again, even without a hanging order currently in place.

- Just like @TheHolyRoger mentioned, the issue with hanging orders failing to recreate the order is still happening on the latest commit.

<img width="1423" alt="Screen Shot 2021-06-09 at 6 56 08 PM" src="https://user-images.githubusercontent.com/64377055/121346710-077f7d00-c959-11eb-853d-b764bb9930ca.png">

- I noticed that sometimes there are delays before all the orders are successfully cancelled. See sample screenshots:

<img width="1421" alt="Screen Shot 2021-06-09 at 6 40 50 PM" src="https://user-images.githubusercontent.com/64377055/121347285-968c9500-c959-11eb-8e22-279f569c7b38.png">

<img width="1421" alt="Screen Shot 2021-06-09 at 6 41 30 PM" src="https://user-images.githubusercontent.com/64377055/121347314-a0ae9380-c959-11eb-965c-f3469b19752d.png">

- I got this error once after stopping the strategy. Unfortunately I wasn't able to reproduce it even after several attempts.
<img width="1421" alt="Screen Shot 2021-06-09 at 6 53 50 PM" src="https://user-images.githubusercontent.com/64377055/121347522-e10e1180-c959-11eb-83d6-6bf3d6622d1f.png">

```
2021-06-09 18:53:15,009 - 5084 - hummingbot.connector.exchange.gate_io.gate_io_exchange - ERROR - cannot convert float NaN to integer
Traceback (most recent call last):
  File "/Users/reiniersalazar/hummingbot/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 567, in _status_polling_loop
    await safe_gather(
  File "/Users/reiniersalazar/hummingbot/hummingbot/core/utils/async_utils.py", line 22, in safe_gather
    return await asyncio.gather(*args, **kwargs)
  File "/Users/reiniersalazar/hummingbot/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 596, in _update_order_status
    current_tick = int(self.current_timestamp / Constants.UPDATE_ORDER_STATUS_INTERVAL)
ValueError: cannot convert float NaN to integer
2021-06-09 18:53:15,016 - 5084 - hummingbot.connector.exchange.gate_io.gate_io_exchange - NETWORK - Unexpected error while fetching account updates.
Traceback (most recent call last):
  File "/Users/reiniersalazar/hummingbot/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 567, in _status_polling_loop
    await safe_gather(
  File "/Users/reiniersalazar/hummingbot/hummingbot/core/utils/async_utils.py", line 22, in safe_gather
    return await asyncio.gather(*args, **kwargs)
  File "/Users/reiniersalazar/hummingbot/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 596, in _update_order_status
    current_tick = int(self.current_timestamp / Constants.UPDATE_ORDER_STATUS_INTERVAL)
ValueError: cannot convert float NaN to integer
2021-06-09 18:53:15,017 - 5084 - hummingbot.connector.exchange.gate_io.gate_io_exchange - WARNING - Could not fetch account updates from gate_io. Check API key and network connection.
```

Logs and config: [gate_io_latest.zip](https://github.com/CoinAlpha/hummingbot/files/6623432/gate_io_latest.zip)

TheHolyRoger commented:
@rxlxrxsx thanks for that, I know why that error is, unlikely race condition but will catch it

Nullably commented on hummingbot/connector/exchange/gate_io/gate_io_exchange.py:
```diff
@@ -0,0 +1,923 @@
+        """
+
+        # Call Update balances on every message to catch order create, fill and cancel - websocket doesn't support this.
+        safe_ensure_future(self._update_balances())
```
I don't think it's a good idea to call update_balances here as it might interfere with the internal balance accounting mechanism I told you about.

TheHolyRoger replied:
@Nullably is update balances automatically called when switching real time off then?

Nullably replied:
No, the available balances are calculated by accounting for current inflight orders.  
When you use `safe_ensure_future` you don't get the execution immediate, you just start the async API call process. And you want to be careful with API rate limit as well.

TheHolyRoger replied:
I've moved the update balance call to the end of the function: https://github.com/CoinAlpha/hummingbot/pull/3484/files?file-filters%5B%5D=.py#diff-589585025595c725ebc393ef67f553c926f35a9409829787b86536bc890e808aR704

Surely `safe_ensure_future` is what I want to use though as I don't want to block order creation returning until the REST balance has updated do I?

TheHolyRoger commented:
Not sure why the test is failing... Think coverage has just been added to the workflow?

RC-13 commented:
- After running the bot for a while, I didn't get the error previously reported when stopping the bot. However, there are instances that the `cancel_all` function fails. Resulting in the following error and leaving the orders in the exchange:
```python
2021-06-11 02:13:07,059 - 13637 - hummingbot.connector.exchange.gate_io.gate_io_exchange - NETWORK - Unexpected error cancelling orders.                                                     
Traceback (most recent call last):                                                                                                                                                          
  File "/home/ubuntu/connector-gate-io/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 822, in cancel_all                                                                   
    cancellation_results = await safe_gather(*tasks, return_exceptions=False)                                                                                                                
  File "/home/ubuntu/connector-gate-io/hummingbot/core/utils/async_utils.py", line 22, in safe_gather                                                                                        
    return await asyncio.gather(*args, **kwargs)                                                                                                                                            
asyncio.exceptions.CancelledError                                                                                                                                                          
                                                                                                                                                                                          
During handling of the above exception, another exception occurred:                                                                                                                          
                                                                                                                                                                                             
Traceback (most recent call last):                                                                                                                                                           
  File "/home/ubuntu/connector-gate-io/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 822, in cancel_all                                                                   
    cancellation_results = await safe_gather(*tasks, return_exceptions=False)                                                                                                                
  File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/async_timeout/__init__.py", line 55, in __aexit__                                                                
    self._do_exit(exc_type)                                                                                                                                                                 
  File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/async_timeout/__init__.py", line 92, in _do_exit                                                                 
    raise asyncio.TimeoutError                                                                                                                                                             
asyncio.exceptions.TimeoutError                                                                                                                                                              
2021-06-11 02:13:07,062 - 13637 - hummingbot.connector.exchange.gate_io.gate_io_exchange - WARNING - Failed to cancel all orders on gate_io. Check API key and network connection.
```
![gate_io_cancel_all](https://user-images.githubusercontent.com/57189990/121621737-56372f00-ca9f-11eb-93a4-c396bcd9658a.png)

- Also when using multiple order levels, there are random instances that the network suddenly drops however it doesn't cause an issue to the active orders:
![image](https://user-images.githubusercontent.com/57189990/121622041-d9f11b80-ca9f-11eb-82ea-5203c7fa2c7f.png)

Nullably commented:
> Not sure why the test is failing... Think coverage has just been added to the workflow?
@TheHolyRoger 
Yep, but we're going to exclude connectors from the coverage, should be done by tomorrow.

TheHolyRoger commented:
I've pushed a fix to increase the cancel_all timeout, see if that's better?

@Nullably I'm seeing weird things going on with these logs, if you look below you'll see a few instances where an order is tracked, then cancelled, THEN created.

A - I'm not sure why the order is cancelled before creation?

B - We're running into timeout errors because of this, since often we get an asyncio timeout during order cancellation, whilst waiting for the order to actually be created...?

JeremyKono commented:
Upon testing, we have received this error on starting the strategy. This is on the latest commit.

Steps to reproduce: 
- Run a pure market making bot
- Choose gate-io as the connector
- The error will show up immediately after starting the bot
```
2021-06-11 11:04:30,942 - 1 - hummingbot.client.hummingbot_application - INFO - start command initiated.
2021-06-11 11:04:31,263 - 1 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Network status has changed to NetworkStatus.CONNECTED. Starting networking...
2021-06-11 11:04:31,577 - 1 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task: 'bids'
Traceback (most recent call last):
  File "/home/hummingbot/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/home/hummingbot/hummingbot/core/data_type/order_book_tracker.py", line 178, in _init_order_books
    self._order_books[trading_pair] = await self._data_source.get_new_order_book(trading_pair)
  File "/home/hummingbot/hummingbot/connector/exchange/gate_io/gate_io_api_order_book_data_source.py", line 88, in get_new_order_book
    bids, asks = active_order_tracker.convert_snapshot_message_to_order_book_row(snapshot_msg)
  File "hummingbot/connector/exchange/gate_io/gate_io_active_order_tracker.pyx", line 146, in hummingbot.connector.exchange.gate_io.gate_io_active_order_tracker.GateIoActiveOrderTracker.convert_snapshot_message_to_order_book_row
    np_bids, np_asks = self.c_convert_snapshot_message_to_np_arrays(message)
  File "hummingbot/connector/exchange/gate_io/gate_io_active_order_tracker.pyx", line 112, in hummingbot.connector.exchange.gate_io.gate_io_active_order_tracker.GateIoActiveOrderTracker.c_convert_snapshot_message_to_np_arrays
    for snapshot_orders, active_orders in [(content["bids"], self._active_bids), (content["asks"], self._active_asks)]:
KeyError: 'bids'
```

TheHolyRoger commented:
> Upon testing, we have received this error on starting the strategy. This is on the latest commit.
> 
> Steps to reproduce:
> 
> * Run a pure market making bot
> * Choose gate-io as the connector
> * The error will show up immediately after starting the bot
> 
> ```
> 2021-06-11 11:04:30,942 - 1 - hummingbot.client.hummingbot_application - INFO - start command initiated.
> 2021-06-11 11:04:31,263 - 1 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Network status has changed to NetworkStatus.CONNECTED. Starting networking...
> 2021-06-11 11:04:31,577 - 1 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task: 'bids'
> Traceback (most recent call last):
>   File "/home/hummingbot/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
>     return await c
>   File "/home/hummingbot/hummingbot/core/data_type/order_book_tracker.py", line 178, in _init_order_books
>     self._order_books[trading_pair] = await self._data_source.get_new_order_book(trading_pair)
>   File "/home/hummingbot/hummingbot/connector/exchange/gate_io/gate_io_api_order_book_data_source.py", line 88, in get_new_order_book
>     bids, asks = active_order_tracker.convert_snapshot_message_to_order_book_row(snapshot_msg)
>   File "hummingbot/connector/exchange/gate_io/gate_io_active_order_tracker.pyx", line 146, in hummingbot.connector.exchange.gate_io.gate_io_active_order_tracker.GateIoActiveOrderTracker.convert_snapshot_message_to_order_book_row
>     np_bids, np_asks = self.c_convert_snapshot_message_to_np_arrays(message)
>   File "hummingbot/connector/exchange/gate_io/gate_io_active_order_tracker.pyx", line 112, in hummingbot.connector.exchange.gate_io.gate_io_active_order_tracker.GateIoActiveOrderTracker.c_convert_snapshot_message_to_np_arrays
>     for snapshot_orders, active_orders in [(content["bids"], self._active_bids), (content["asks"], self._active_asks)]:
> KeyError: 'bids'
> ```

@JeremyKono could you tell me which pair this is with please?

TheHolyRoger commented:
@JeremyKono found and fixed that one, API changed... :)

JeremyKono commented:
It's not specific to any pair as I have tried multiple pairs.

TheHolyRoger commented:
> It's not specific to any pair as I have tried multiple pairs.

thanks man should be fixed now

RC-13 commented:
Confirming fix for the issue reported when starting a strategy using gate.io. Will post back if we encounter a new error.
>Upon testing, we have received this error on starting the strategy. This is on the latest commit.

rxlxrxsx commented:
- This is still an issue on the latest commit.

> Hanging orders refreshed by max order age is not recreated due to the delay in balance update.

<img width="1422" alt="Screen Shot 2021-06-13 at 2 52 24 PM" src="https://user-images.githubusercontent.com/64377055/121799884-5aab4580-cc61-11eb-955e-20ef91f50c26.png">

- This issue is happening again on the latest commit.

> There are instances that the bot skips an order cycle and won't create orders during those period.

<img width="1423" alt="Screen Shot 2021-06-13 at 2 58 18 PM" src="https://user-images.githubusercontent.com/64377055/121799991-f50b8900-cc61-11eb-86be-a5acc2f8a71f.png">

- There are also instances wherein there's a slight delay on orders being created. See screenshot below:

<img width="1422" alt="Screen Shot 2021-06-13 at 3 32 36 PM" src="https://user-images.githubusercontent.com/64377055/121800033-22583700-cc62-11eb-96e7-61ebaddb275f.png">

Logs and config: [gateio.zip](https://github.com/CoinAlpha/hummingbot/files/6643549/gateio.zip)

Nullably commented:
@rxlxrxsx 
After looking into the code on pure market making strategy, it looks like a possible existing bug on the max order age feature. Does this feature (regardless if it is for hanging orders or not) work on connectors where balance update is not realtime, e.g. Coinbase Pro or Kraken?

rxlxrxsx commented:
> @rxlxrxsx
> After looking into the code on pure market making strategy, it looks like a possible existing bug on the max order age feature. Does this feature (regardless if it is for hanging orders or not) work on connectors where balance update is not realtime, e.g. Coinbase Pro or Kraken?

@Nullably I'll take a look into this one.

Nullably commented:
@TheHolyRoger 
> @Nullably I'm seeing weird things going on with these logs, if you look below you'll see a few instances where an order is tracked, then cancelled, THEN created.

This is expected for `max_order_age` feature where the strategy replaces exiting orders with new ones (cancels then creates) of the same order amount and price if the order is older than max_order_age. Our testers set this to be a short period, e.g. 30 s, for testing purpose.

There maybe a bug I mentioned above where the bot doesn't wait for the cancel confirmation before place new orders.

TheHolyRoger commented:
> @TheHolyRoger
> 
> > @Nullably I'm seeing weird things going on with these logs, if you look below you'll see a few instances where an order is tracked, then cancelled, THEN created.
> 
> This is expected for `max_order_age` feature where the strategy replaces exiting orders with new ones (cancels then creates) of the same order amount and price if the order is older than max_order_age. Our testers set this to be a short period, e.g. 30 s, for testing purpose.
> 
> There maybe a bug I mentioned above where the bot doesn't wait for the cancel confirmation before placing new orders.

@Nullably see this section of the log, look at order `t-HBOT-B-BCUT1623397913004734` - I added an additional log line for `Tracking` when an order is first tracked

```
2021-06-11 08:51:53,007 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Tracking t-HBOT-B-BCUT1623397913004699.
2021-06-11 08:51:53,008 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Tracking t-HBOT-B-BCUT1623397913004734.
2021-06-11 08:51:53,008 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Tracking t-HBOT-B-BCUT1623397913004752.
2021-06-11 08:51:53,008 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Tracking t-HBOT-B-BCUT1623397913004768.
2021-06-11 08:51:53,008 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Tracking t-HBOT-B-BCUT1623397913004782.
2021-06-11 08:51:53,009 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Tracking t-HBOT-B-BCUT1623397913004797.
2021-06-11 08:51:53,256 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Successfully cancelled order t-HBOT-B-BCUT1623397872001441 on gate_io.
2021-06-11 08:51:53,264 - 59744 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623397913.0, "order_id": "t-HBOT-B-BCUT1623397872001441", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "gate_io"}
2021-06-11 08:51:54,109 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Successfully cancelled order t-HBOT-B-BCUT1623397872001531 on gate_io.
2021-06-11 08:51:54,118 - 59744 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623397914.0, "order_id": "t-HBOT-B-BCUT1623397872001531", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "gate_io"}
2021-06-11 08:51:54,119 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Successfully cancelled order t-HBOT-B-BCUT1623397872001547 on gate_io.
2021-06-11 08:51:54,125 - 59744 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623397914.0, "order_id": "t-HBOT-B-BCUT1623397872001547", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "gate_io"}
2021-06-11 08:51:54,163 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Successfully cancelled order t-HBOT-B-BCUT1623397872001510 on gate_io.
2021-06-11 08:51:54,170 - 59744 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623397914.0, "order_id": "t-HBOT-B-BCUT1623397872001510", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "gate_io"}
2021-06-11 08:51:57,000 - 59744 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (BTC-USDT) Cancelling the limit order t-HBOT-B-BCUT1623397913004699. [clock=2021-06-11 07:51:57+00:00]
2021-06-11 08:51:57,001 - 59744 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (BTC-USDT) Cancelling the limit order t-HBOT-B-BCUT1623397913004734. [clock=2021-06-11 07:51:57+00:00]
2021-06-11 08:51:57,001 - 59744 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (BTC-USDT) Cancelling the limit order t-HBOT-B-BCUT1623397913004752. [clock=2021-06-11 07:51:57+00:00]
2021-06-11 08:51:57,001 - 59744 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (BTC-USDT) Cancelling the limit order t-HBOT-B-BCUT1623397913004768. [clock=2021-06-11 07:51:57+00:00]
2021-06-11 08:51:57,001 - 59744 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (BTC-USDT) Cancelling the limit order t-HBOT-B-BCUT1623397913004782. [clock=2021-06-11 07:51:57+00:00]
2021-06-11 08:51:57,001 - 59744 - hummingbot.strategy.pure_market_making.pure_market_making - INFO - (BTC-USDT) Cancelling the limit order t-HBOT-B-BCUT1623397913004797. [clock=2021-06-11 07:51:57+00:00]
2021-06-11 08:51:57,184 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Created LIMIT BUY order t-HBOT-B-BCUT1623397913004734 for 0.0001 BTC-USDT.
2021-06-11 08:51:57,190 - 59744 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623397917.0, "type": "OrderType.LIMIT", "trading_pair": "BTC-USDT", "amount": "0.0001", "price": "35492.49", "order_id": "t-HBOT-B-BCUT1623397913004734", "exchange_order_id": "54062735552", "leverage": 1, "position": "NILL", "event_name": "BuyOrderCreatedEvent", "event_source": "gate_io"}
2021-06-11 08:51:57,229 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Successfully cancelled order t-HBOT-B-BCUT1623397872001573 on gate_io.
2021-06-11 08:51:57,236 - 59744 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623397917.0, "order_id": "t-HBOT-B-BCUT1623397872001573", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "gate_io"}
2021-06-11 08:51:57,237 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Created LIMIT BUY order t-HBOT-B-BCUT1623397913004699 for 0.0001 BTC-USDT.
2021-06-11 08:51:57,242 - 59744 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623397917.0, "type": "OrderType.LIMIT", "trading_pair": "BTC-USDT", "amount": "0.0001", "price": "36601.63", "order_id": "t-HBOT-B-BCUT1623397913004699", "exchange_order_id": "54062735550", "leverage": 1, "position": "NILL", "event_name": "BuyOrderCreatedEvent", "event_source": "gate_io"}
2021-06-11 08:51:57,242 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Successfully cancelled order t-HBOT-B-BCUT1623397872001561 on gate_io.
2021-06-11 08:51:57,247 - 59744 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623397917.0, "order_id": "t-HBOT-B-BCUT1623397872001561", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "gate_io"}
2021-06-11 08:51:58,765 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Successfully cancelled order t-HBOT-B-BCUT1623397913004699 on gate_io.
2021-06-11 08:51:58,771 - 59744 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623397918.0, "order_id": "t-HBOT-B-BCUT1623397913004699", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "gate_io"}
2021-06-11 08:52:00,200 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_order_book_tracker - DEBUG - Diff messages processed: 502, rejected: 0
2021-06-11 08:52:00,201 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_order_book_tracker - DEBUG - Processed 453 order book diffs for BTC-USDT.
2021-06-11 08:52:00,334 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_order_book_tracker - DEBUG - Trade messages processed: 95, rejected: 0
2021-06-11 08:52:01,108 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Created LIMIT BUY order t-HBOT-B-BCUT1623397913004782 for 0.0001 BTC-USDT.
2021-06-11 08:52:01,115 - 59744 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623397921.0, "type": "OrderType.LIMIT", "trading_pair": "BTC-USDT", "amount": "0.0001", "price": "32165.07", "order_id": "t-HBOT-B-BCUT1623397913004782", "exchange_order_id": "54062742377", "leverage": 1, "position": "NILL", "event_name": "BuyOrderCreatedEvent", "event_source": "gate_io"}
2021-06-11 08:52:01,217 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Created LIMIT BUY order t-HBOT-B-BCUT1623397913004768 for 0.0001 BTC-USDT.
2021-06-11 08:52:01,223 - 59744 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623397921.0, "type": "OrderType.LIMIT", "trading_pair": "BTC-USDT", "amount": "0.0001", "price": "33274.21", "order_id": "t-HBOT-B-BCUT1623397913004768", "exchange_order_id": "54062742368", "leverage": 1, "position": "NILL", "event_name": "BuyOrderCreatedEvent", "event_source": "gate_io"}
2021-06-11 08:52:01,224 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Created LIMIT BUY order t-HBOT-B-BCUT1623397913004752 for 0.0001 BTC-USDT.
2021-06-11 08:52:01,229 - 59744 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623397921.0, "type": "OrderType.LIMIT", "trading_pair": "BTC-USDT", "amount": "0.0001", "price": "34383.35", "order_id": "t-HBOT-B-BCUT1623397913004752", "exchange_order_id": "54062742379", "leverage": 1, "position": "NILL", "event_name": "BuyOrderCreatedEvent", "event_source": "gate_io"}
2021-06-11 08:52:05,114 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Created LIMIT BUY order t-HBOT-B-BCUT1623397913004797 for 0.0001 BTC-USDT.
2021-06-11 08:52:05,122 - 59744 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623397925.0, "type": "OrderType.LIMIT", "trading_pair": "BTC-USDT", "amount": "0.0001", "price": "31055.93", "order_id": "t-HBOT-B-BCUT1623397913004797", "exchange_order_id": "54062752284", "leverage": 1, "position": "NILL", "event_name": "BuyOrderCreatedEvent", "event_source": "gate_io"}
2021-06-11 08:52:09,044 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Successfully cancelled order t-HBOT-B-BCUT1623397913004734 on gate_io.
2021-06-11 08:52:09,052 - 59744 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623397929.0, "order_id": "t-HBOT-B-BCUT1623397913004734", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "gate_io"}
2021-06-11 08:52:13,038 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Successfully cancelled order t-HBOT-B-BCUT1623397913004768 on gate_io.
2021-06-11 08:52:13,044 - 59744 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623397933.0, "order_id": "t-HBOT-B-BCUT1623397913004768", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "gate_io"}
2021-06-11 08:52:13,044 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Successfully cancelled order t-HBOT-B-BCUT1623397913004752 on gate_io.
2021-06-11 08:52:13,049 - 59744 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623397933.0, "order_id": "t-HBOT-B-BCUT1623397913004752", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "gate_io"}
2021-06-11 08:52:13,050 - 59744 - hummingbot.connector.exchange.gate_io.gate_io_exchange - INFO - Successfully cancelled order t-HBOT-B-BCUT1623397913004782 on gate_io.
2021-06-11 08:52:13,056 - 59744 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1623397933.0, "order_id": "t-HBOT-B-BCUT1623397913004782", "exchange_order_id": null, "event_name": "OrderCancelledEvent", "event_source": "gate_io"}
```

Order `t-HBOT-B-BCUT1623397913004734` is cancelled before the **"created"** event has even been received which seems wrong to me?

Nullably commented:
@TheHolyRoger 
> Order t-HBOT-B-BCUT1623397913004734 is cancelled before the "created" event has even been received which seems wrong to me?

Sorry I overlooked that. Yeah that seems wrong to me as well.
This is a problem in the strategy where it considers any submitted orders to be active orders, including ones that have not got API response. In your case `Order t-HBOT-B-BCUT1623397913004734` took about 4 seconds

rxlxrxsx commented:
- When gate.io is used as a taker market on cross exchange and arbitrage strategies, the strategy fails to execute an order and will throw this error upon an opportunity:

<img width="1423" alt="Screen Shot 2021-06-21 at 3 38 25 PM" src="https://user-images.githubusercontent.com/64377055/122725157-4988b700-d2a7-11eb-9e03-2806bb4794a7.png">

```
2021-06-21 15:29:21,911 - 7983 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task: Unsupported order type: OrderType.MARKET
Traceback (most recent call last):
  File "/Users/reiniersalazar/hummingbot4/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/Users/reiniersalazar/hummingbot4/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 438, in _create_order
    raise Exception(f"Unsupported order type: {order_type}")
Exception: Unsupported order type: OrderType.MARKET
```

Logs and config: [gate_io_taker.zip](https://github.com/CoinAlpha/hummingbot/files/6684976/gate_io_taker.zip)

TheHolyRoger commented:
@rxlxrxsx thanks for that, I can see the problem on line 438, silly mistake on my part! Will try and push a fix later today or tomorrow

TheHolyRoger commented:
@rxlxrxsx this should be sorted now!

RC-13 commented:
Confirming that `Market` order no longer gets the error reported above, however, taker order (gate.io) fails to be submitted. This happens when using cross-exchange or arbitrage strategy and with the following error:
```python
2021-06-28 08:24:32,459 - 3101 - hummingbot.connector.exchange.gate_io.gate_io_exchange - NETWORK - Error submitting SELL MARKET order to gate_io for 0.0003 BTC-USDT NaN - Request body not conforming to schema.
Traceback (most recent call last):
  File "/Users/clint_g/Documents/Github-Repo/gate-io/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 454, in _create_order
    order_result = await self._api_request("POST", Constants.ENDPOINT["ORDER_CREATE"], api_params, True)
  File "/Users/clint_g/Documents/Github-Repo/gate-io/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 359, in _api_request
    raise GateIoAPIError(parsed_response)
hummingbot.connector.exchange.gate_io.gate_io_utils.GateIoAPIError: {'label': 'INVALID_REQUEST_BODY', 'message': 'Request body not conforming to schema'}
```
![image](https://user-images.githubusercontent.com/57189990/123566810-3455e000-d7f3-11eb-8394-85d799eccc93.png)
[logs_conf_arb_1.log](https://github.com/CoinAlpha/hummingbot/files/6722688/logs_conf_arb_1.log)
[logs_conf_xemm_2.log](https://github.com/CoinAlpha/hummingbot/files/6722689/logs_conf_xemm_2.log)

TheHolyRoger commented:
@rxlxrxsx sorry to mess you about! Looks like gate.io may not support market orders via the API. Just confirming before I update the code

TheHolyRoger commented:
@Nullably @rxlxrxsx Gate.io doesn't support Market orders, they do support something called time in force, not sure if that's any use?

https://www.gate.io/docs/apiv4/en/index.html#create-an-order

Otherwise i'll disable market orders in an update

phbrgnomo commented:
> @Nullably @rxlxrxsx Gate.io doesn't support Market orders, they do support something called time in force, not sure if that's any use?
> 
> https://www.gate.io/docs/apiv4/en/index.html#create-an-order
> 
> Otherwise i'll disable market orders in an update

I think `time_in_force` is what the instructions what to do after the order is created:
`gtc` - order stay up until is cancelled by the user
`ioc` - order is cancelled if it isn't immediately filled. if it is a limit order, it would work as a taker order, being canceled if there is no match. (But it mentions "take only on the API doc)

Maybe it would work creating a limit order with `time_in_force = ioc` would have the same effect?

rxlxrxsx commented:
Posting issues that has been fixed and is still happening on the latest commit:

- This is now fixed
> Hanging orders refreshed by max order age is not recreated due to the delay in balance update.

- This is still an issue on the latest commit
> There are instances that the bot skips an order cycle and won't create orders during those period.

<img width="1424" alt="Screen Shot 2021-06-30 at 12 06 11 PM" src="https://user-images.githubusercontent.com/64377055/123907351-5a70b100-d9a8-11eb-87a7-ba023eeb90c6.png">

- This is still an issue on the latest commit, I also noticed that this is easier to reproduce on multiple order levels
> There are also instances wherein there's a slight delay on orders being created.

<img width="1422" alt="Screen Shot 2021-06-30 at 12 07 23 PM" src="https://user-images.githubusercontent.com/64377055/123907424-73796200-d9a8-11eb-93c9-f6cab5a23df6.png">

<img width="1422" alt="Screen Shot 2021-06-30 at 12 55 54 PM" src="https://user-images.githubusercontent.com/64377055/123907479-8ee46d00-d9a8-11eb-9829-d14491488b7b.png">

- I set the order levels to 10. As you can see on the screenshot above this, it can only create 4-5 orders based on the available balance but there are instances that it creates 10 orders on each side but was non existent on the exchange. I also noticed that errors started appearing when using multiple order levels

<img width="571" alt="Screen Shot 2021-06-30 at 12 47 29 PM" src="https://user-images.githubusercontent.com/64377055/123907736-f7334e80-d9a8-11eb-8cd4-32c6abcac3e5.png">

```python
2021-06-30 12:44:42,049 - 4695 - hummingbot.connector.exchange.gate_io.gate_io_exchange - NETWORK - Error submitting BUY LIMIT order to gate_io for 0.0003 BTC-USDT 33076.58 - Not enough balance.
Traceback (most recent call last):
  File "/Users/reiniersalazar/hummingbot4/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 454, in _create_order
    order_result = await self._api_request("POST", Constants.ENDPOINT["ORDER_CREATE"], api_params, True)
  File "/Users/reiniersalazar/hummingbot4/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 359, in _api_request
    raise GateIoAPIError(parsed_response)
hummingbot.connector.exchange.gate_io.gate_io_utils.GateIoAPIError: {'label': 'BALANCE_NOT_ENOUGH', 'message': 'Not enough balance'}
```

```python
2021-06-30 12:46:35,004 - 4695 - hummingbot.connector.exchange.gate_io.gate_io_exchange - ERROR - 
Traceback (most recent call last):
  File "hummingbot/connector/in_flight_order_base.pyx", line 84, in hummingbot.connector.in_flight_order_base.InFlightOrderBase.get_exchange_order_id
    await self.exchange_order_id_update_event.wait()
  File "/Users/reiniersalazar/opt/anaconda3/envs/hummingbot/lib/python3.8/asyncio/locks.py", line 309, in wait
    await fut
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/reiniersalazar/hummingbot4/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 575, in _status_polling_loop
    await safe_gather(
  File "/Users/reiniersalazar/hummingbot4/hummingbot/core/utils/async_utils.py", line 22, in safe_gather
    return await asyncio.gather(*args, **kwargs)
  File "/Users/reiniersalazar/hummingbot4/hummingbot/connector/exchange/gate_io/gate_io_exchange.py", line 612, in _update_order_status
    exchange_order_id = await tracked_order.get_exchange_order_id()
  File "hummingbot/connector/in_flight_order_base.pyx", line 83, in get_exchange_order_id
    async with timeout(10):
  File "/Users/reiniersalazar/opt/anaconda3/envs/hummingbot/lib/python3.8/site-packages/async_timeout/__init__.py", line 55, in __aexit__
    self._do_exit(exc_type)
  File "/Users/reiniersalazar/opt/anaconda3/envs/hummingbot/lib/python3.8/site-packages/async_timeout/__init__.py", line 92, in _do_exit
    raise asyncio.TimeoutError
asyncio.exceptions.TimeoutError
```

Logs and config: [gate_io_latest.zip](https://github.com/CoinAlpha/hummingbot/files/6738368/gate_io_latest.zip)

TheHolyRoger commented:
@rxlxrxsx Looking at these issues now.

@phbrgnomo Could use a little more direction on this please :) I don't really understand what these params mean and how to make "market" orders, or if it's acceptable for the arb strategies....

TheHolyRoger commented:
@rxlxrxsx I'm struggling to reproduce the issue of ghost orders being created in hummingbot, but I've added an order status check which I think might have been the problem.

If you encounter more of these can you please manually get the order status of the `exchange_order_id` it shows in the logs for the ghost orders?

They *should* disappear at most after 3 status checks show the order as not found.

I've also added a special test file to retrieve order status - you need to do this while the bot is running and thinks the order is "open" - otherwise as soon as the order is cancelled it returns not found.

Set the usual GATE_IO api environment variables then also these ones:
```bash
export TEST_ORDER_ID=58717915545 && export TEST_TRADING_PAIR=BTC_USDT

nosetests -d -v --nocapture -x test/connector/exchange/gate_io/test_gate_io_order_status.py
```

Still awaiting more info to handle **market** orders so don't test arb yet

rxlxrxsx commented:
- This issue seems to have been fixed on the latest commit as I'm not able to replicate it anymore.
> @rxlxrxsx I'm struggling to reproduce the issue of ghost orders being created in hummingbot, but I've added an order status check which I think might have been the problem.

- I can see that this only now happens on multiple order levels.
> There are also instances wherein there's a slight delay on orders being created.

<img width="1421" alt="Screen Shot 2021-07-01 at 9 58 37 AM" src="https://user-images.githubusercontent.com/64377055/124063773-31aef100-da66-11eb-90b9-be552a9ec6cb.png">

- I also noticed that I frequently get disconnected when on multiple order levels, I wasn't able to encounter this while on a single order level.

<img width="1421" alt="Screen Shot 2021-07-01 at 9 57 11 AM" src="https://user-images.githubusercontent.com/64377055/124063911-6b7ff780-da66-11eb-9c71-1d6c1da8e2b8.png">

- It seems that the inconsistent order creation issue and available balances not updating is happening again (I think this one is related to the last bullet point).

<img width="1423" alt="Screen Shot 2021-07-01 at 10 36 55 AM" src="https://user-images.githubusercontent.com/64377055/124064230-f19c3e00-da66-11eb-983f-0c877e4f4bdd.png">

<img width="1422" alt="Screen Shot 2021-07-01 at 10 04 37 AM" src="https://user-images.githubusercontent.com/64377055/124064290-08429500-da67-11eb-8498-b8c3c0381790.png">

- Also on this screenshot you'll notice on the log pane that the orders are still being created however when you check on status the orders are already there, I can confirm that the orders will only start appearing on the exchange when it says on the log pane `Created LIMIT BUY/SELL order...`, so technically those orders are being open less than the order refresh time set.

<img width="1423" alt="Screen Shot 2021-07-01 at 12 36 19 PM" src="https://user-images.githubusercontent.com/64377055/124065350-34f7ac00-da69-11eb-8d21-6736e99fbe45.png">

<img width="762" alt="Screen Shot 2021-07-01 at 12 36 34 PM" src="https://user-images.githubusercontent.com/64377055/124065364-3c1eba00-da69-11eb-9bda-8d0661885cc6.png">

Logs and config: [gate-io.zip](https://github.com/CoinAlpha/hummingbot/files/6745429/gate-io.zip)

Nullably commented:
@TheHolyRoger 
You can add the below function in `gate_io_exchange`, what it does is it tells strategy to use LIMIT order type when taker order is expected. 
```python
def get_taker_order_type(self):
        return OrderType.LIMIT
```

Also, I believe you shouldn't make a call to `update_balances` on every process_order_message, this might be what causes lagging that the tester experienced.

TheHolyRoger commented:
> @TheHolyRoger
> You can add the below function in `gate_io_exchange`, what it does is it tells strategy to use LIMIT order type when taker order is expected.
> 
> ```python
> def get_taker_order_type(self):
>         return OrderType.LIMIT
> ```
> 
> You can also remove OrderType.MARKET from its supported order types.
> 
> Also, I believe you shouldn't make a call to `update_balances` on every process_order_message, this might be what causes lagging that the tester experienced.

@Nullably where do you think I should call update balances? Since there's no real time updates... 

Thank you will make that limit order adjustment

TheHolyRoger commented:
I've updated this with a little queue mechanism for the balance updates and a couple of other things to hopefully help this lag but keep balance updates quick.

@rxlxrxsx bear in mind that gate.io API is a tad laggy in general ;)

I've also added a logging line for the purposes of catching unexpected order types for now and spitting out the full order JSON, this might trigger if a limit order is immediately matched so would be nice to see how the gate.io API handles these, I haven't managed to trigger anything myself yet

Nullably commented:
@TheHolyRoger 
> @Nullably where do you think I should call update balances? Since there's no real time updates...

Does it not work well if you just have the call on the status polling loop?

PtrckM commented:
Issues fix on the latest commit
      - hanging order not recreated = fix
      - bot skips an order cycle and won't create orders = fix
      - delay / skip order creation on multiple order level with 10 second refresh time = fix

Order creation
      - slight delay on orders being created = inconsistent (not always)

We will run long term test bot for monitoring.

dennisocana commented:
I think the connector is pretty much usable and we can merge this to `development` so it will be included in version 0.40. We should create a Github ticket about the random issue where orders are not created at certain cycles. We can push the fix to pre-prod (staging) next week before we release on Friday, July 9th.

TheHolyRoger commented:
> @TheHolyRoger
> 
> > @Nullably where do you think I should call update balances? Since there's no real time updates...
> 
> Does it not work well if you just have the call on the status polling loop?

Unfortunately not, unless it's a really short interval.

But this seems much happier now, it will only call one balance update at a time and will always catch the first and last calls using order updates as a trigger

Nullably reviewed (approved):
LGTM

rapcmia reviewed (approved):
Tested and confirmed working by QA
